File: clients/android/app/src/test/java/com/kurisu/assistant/domain/version/VersionParityTest.kt
package com.kurisu.assistant.domain.version

import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Test

/**
 * About says when the app and the backend are different releases (#257); the
 * rules mirror the desktop's `versionParity.test.ts`.
 */
class VersionParityTest {

    @Test
    fun `keeps major minor patch and drops the rest`() {
        assertEquals("0.7.0", VersionParity.releaseCore("0.7.0"))
        assertEquals("0.7.0", VersionParity.releaseCore("v0.7.0"))
        assertEquals("0.7.0", VersionParity.releaseCore("0.7.0-dev-a1b2c3d-dirty"))
        assertEquals("0.7.0", VersionParity.releaseCore("0.7.0+42"))
    }

    @Test
    fun `is null for nothing and for strings with no release in them`() {
        assertNull(VersionParity.releaseCore(null))
        assertNull(VersionParity.releaseCore(""))
        assertNull(VersionParity.releaseCore("?"))
        assertNull(VersionParity.releaseCore("0.7"))
    }

    @Test
    fun `says nothing when both sides are the same release`() {
        assertNull(VersionParity.mismatchSentence("0.7.0", "0.7.0"))
        assertNull(VersionParity.mismatchSentence("0.7.0-dev-abc1234", "v0.7.0"))
    }

    @Test
    fun `names both numbers when they differ`() {
        assertEquals(
            "This app is v0.7.0; the backend is v0.6.0 — update whichever is behind.",
            VersionParity.mismatchSentence("0.7.0", "0.6.0"),
        )
    }

    @Test
    fun `says nothing when either side is unknown`() {
        assertNull(VersionParity.mismatchSentence(null, "0.7.0"))
        assertNull(VersionParity.mismatchSentence("0.7.0", null))
        assertNull(VersionParity.mismatchSentence("0.7.0", "?"))
    }
}

File: clients/packages/models/src/versionParity.test.ts
import { describe, expect, it } from 'vitest';
import { releaseCore, versionMismatchSentence } from './versionParity';

describe('releaseCore', () => {
  it('keeps major.minor.patch and drops the rest', () => {
    expect(releaseCore('0.7.0')).toBe('0.7.0');
    expect(releaseCore('v0.7.0')).toBe('0.7.0');
    expect(releaseCore('0.7.0-dev-a1b2c3d-dirty')).toBe('0.7.0');
    expect(releaseCore('0.7.0+42')).toBe('0.7.0');
  });

  it('is null for nothing and for strings with no release in them', () => {
    expect(releaseCore(null)).toBeNull();
    expect(releaseCore(undefined)).toBeNull();
    expect(releaseCore('')).toBeNull();
    expect(releaseCore('unknown')).toBeNull();
    expect(releaseCore('0.7')).toBeNull();
  });
});

describe('versionMismatchSentence', () => {
  it('says nothing when both sides are the same release', () => {
    expect(versionMismatchSentence('0.7.0', '0.7.0')).toBeNull();
    expect(versionMismatchSentence('v0.7.0', '0.7.0-dev-abc1234')).toBeNull();
  });

  it('names both numbers when they differ', () => {
    expect(versionMismatchSentence('0.7.0', '0.6.0')).toBe(
      'This app is v0.7.0; the backend is v0.6.0 — update whichever is behind.',
    );
  });

  it('says nothing when either side is unknown', () => {
    expect(versionMismatchSentence(null, '0.7.0')).toBeNull();
    expect(versionMismatchSentence('0.7.0', null)).toBeNull();
    expect(versionMismatchSentence('0.7.0', 'unknown')).toBeNull();
  });
});
